### DISCUSSION TITLE
Using Function Graph plugin

### DISCUSSION THREAD

**Discussion by treydur** [NationalSecurityAgency/ghidra#6371] [Q&A] [answered]
Hi,
I think the way Ghidra displays function graphs is great; however, I would like to choose my own code blocks to be displayed on the graph.

![image](https://github.com/NationalSecurityAgency/ghidra/assets/88554819/dd2db240-40ba-4511-af92-278263e720f8)

Specifically, I have been creating `GDirectedGraph<CodeBlockVertex, CodeBlockEdge>`'s so I could apply tarjan's algorithm from the package `ghidra.graph.algo`. Then to visualize it I have to convert it into an `AttributedGraph` with `AttributedVertex` and `AttributedEdge` to visualize it as follows.

![image](https://github.com/NationalSecurityAgency/ghidra/assets/88554819/d9adee77-3342-4e3c-b088-f0529ff09ed0)

Is there a way I can choose my own code blocks via script to be in Ghidra's Function Graph?

**Comment by dragonmacher** [comment]
The API currently does not allow for code block selection when graphs are created.  That being said, you can change the source code to do whatever you like. 

The class that generates the graphs using code blocks is `ghidra.app.plugin.core.functiongraph.graph.FunctionGraphFactory`.   The method `createGraph()` is a good place to start.   It calls `createVertices()`, which uses a `BasicBlockModel` to get Code Blocks. 

Assuming you are setup to change Ghidra source code, you can change how `createVertices()` creates code blocks and then the Function Graph will use your changes.

**Comment by treydur** [comment]
As a follow up, since this whole script was based on wanting to visualize ghidra.graph.algo Tarjan's algorithm, is there a different approach to do this?

**Comment by dragonmacher** [comment, marked as answer]
Creating a script to use our Graph APIs is the easiest way to create and render a graph in Ghidra.  The UI provided for scripting is more general and not as feature rich as the more specialized Function Graph.   

If you only need display, then the method you are using is probably a good one.  If you really want the features of the Function Graph, but would also like to change how the blocks are generated, then the more invasive changing of the source code is what you would need to do.
